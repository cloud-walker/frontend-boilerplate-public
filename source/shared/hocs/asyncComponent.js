import React from 'react'

export default getComponent => React.createClass({
  displayName: 'AsyncComponent',
  getInitialState() {
    return {Component: null}
  },
  componentWillMount() {
    if (!this.state.Component) {
      getComponent().then(Component => this.setState({Component}))
    }
  },
  render() {
    const {Component} = this.state

    return Component && <Component {...this.props}/>
  },
})
