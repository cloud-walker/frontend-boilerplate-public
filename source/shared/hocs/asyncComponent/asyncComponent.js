import React from 'react'

/**
 * This higher order component gets a function that produce
 * the promise from the dynamic import, and use it to create
 * a meta component holding the state of the promise.
 * Until the promise is resolved the component will render
 * nothing.
 *
 * @param  {function} getComponent
 * @return {Component}
 */
export default getComponent => React.createClass({
  displayName: 'AsyncComponent',
  getInitialState() {
    return {Component: null}
  },
  componentWillMount() {
    if (!this.state.Component) {
      this.promiseComponent = getComponent().then(Component => {
        this.setState({Component})
        return Component
      })
    }
  },
  render() {
    const {Component} = this.state
    return Component && <Component {...this.props}/>
  },
})
