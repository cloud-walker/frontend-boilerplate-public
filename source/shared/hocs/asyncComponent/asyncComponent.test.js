import React from 'react'
import {shallow} from 'enzyme'
import asyncComponent from './index'

test('asyncComponent', async () => {
  const lazyComponent = () => new Promise(resolve => {
    setTimeout(function () {
      resolve(props => <div {...props}>foo</div>)
    }, 50)
  })
  const AsyncComponent = asyncComponent(lazyComponent)
  const subject = shallow(<AsyncComponent/>)

  /**
   * At first the component should render nothing as the promise is not
   * yet resolved.
   */
  expect(subject).toMatchSnapshot()

  /**
   * Waiting the promiseComponent to resolve the subject should render
   * a different output.
   */
  await subject.instance().promiseComponent
  expect(subject).toMatchSnapshot()

  /**
   * Here we test the passing prop to the dynamic generated Component
   * functionality.
   */
  subject.setProps({
    foo: 'bar',
    baz: 'taz',
  })
  expect(subject).toMatchSnapshot()

  /**
   * This is to achieve 100% branch coverage.
   *
   * If the componentWillMount trigger more than once, the component
   * should not check the promiseComponent again!
   */
  subject.instance().componentWillMount()
})
