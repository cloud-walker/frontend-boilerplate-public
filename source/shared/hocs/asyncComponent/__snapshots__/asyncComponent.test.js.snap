exports[`test asyncComponent 1`] = `<AsyncComponent />`;

exports[`test asyncComponent 2`] = `
<AsyncComponent>
  <Component>
    <div>
      foo
    </div>
  </Component>
</AsyncComponent>
`;

exports[`test asyncComponent 3`] = `
<AsyncComponent
  className="foo"
  style={
    Object {
      "color": "red",
    }
  }>
  <Component
    className="foo"
    style={
      Object {
        "color": "red",
      }
    }>
    <div
      className="foo"
      style={
        Object {
          "color": "red",
        }
      }>
      foo
    </div>
  </Component>
</AsyncComponent>
`;
